x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_installation/perform_rule_installation_handler.ts

```diff
@@ -0,0 +1,145 @@
+import type { KibanaRequest, KibanaResponseFactory } from '@kbn/core/server';
+import { SkipRuleInstallReason } from '../../../../../../common/api/detection_engine/prebuilt_rules';
+import type {
+  PerformRuleInstallationResponseBody,
+  SkippedRuleInstall,
+  PerformRuleInstallationRequestBody,
+} from '../../../../../../common/api/detection_engine/prebuilt_rules';
+import type { SecuritySolutionRequestHandlerContext } from '../../../../../types';
+import { buildSiemResponse } from '../../../routes/utils';
+import { aggregatePrebuiltRuleErrors } from '../../logic/aggregate_prebuilt_rule_errors';
+import { ensureLatestRulesPackageInstalled } from '../../logic/ensure_latest_rules_package_installed';
+import { createPrebuiltRuleAssetsClient } from '../../logic/rule_assets/prebuilt_rule_assets_client';
+import { createPrebuiltRules } from '../../logic/rule_objects/create_prebuilt_rules';
+import { createPrebuiltRuleObjectsClient } from '../../logic/rule_objects/prebuilt_rule_objects_client';
+import { performTimelinesInstallation } from '../../logic/perform_timelines_installation';
+import type { RuleSignatureId, RuleVersion } from '../../../../../../common/api/detection_engine';
+
+export const performRuleInstallationHandler = async (
+  context: SecuritySolutionRequestHandlerContext,
+  request: KibanaRequest<unknown, unknown, PerformRuleInstallationRequestBody>,
+  response: KibanaResponseFactory
+) => {
+  const siemResponse = buildSiemResponse(response);
+
+  try {
+    const ctx = await context.resolve(['core', 'alerting', 'securitySolution']);
+    const config = ctx.securitySolution.getConfig();
+    const soClient = ctx.core.savedObjects.client;
+    const rulesClient = await ctx.alerting.getRulesClient();
+    const detectionRulesClient = ctx.securitySolution.getDetectionRulesClient();
+    const ruleAssetsClient = createPrebuiltRuleAssetsClient(soClient);
+    const ruleObjectsClient = createPrebuiltRuleObjectsClient(rulesClient);
+    const exceptionsListClient = ctx.securitySolution.getExceptionListClient();
+
+    const { mode } = request.body;
+
+    // This will create the endpoint list if it does not exist yet
+    await exceptionsListClient?.createEndpointList();
+
+    // If this API is used directly without hitting any detection engine
+    // pages first, the rules package might be missing.
+    await ensureLatestRulesPackageInstalled(ruleAssetsClient, config, ctx.securitySolution);
+
+    const allLatestVersions = await ruleAssetsClient.fetchLatestVersions();
+    const currentRuleVersions = await ruleObjectsClient.fetchInstalledRuleVersions();
+    const currentRuleVersionsMap = new Map(
+      currentRuleVersions.map((version) => [version.rule_id, version])
+    );
+
+    const allInstallableRules = allLatestVersions.filter((latestVersion) => {
+      const currentVersion = currentRuleVersionsMap.get(latestVersion.rule_id);
+      return !currentVersion;
+    });
```

Comment:
nit: `installableRules` calculation repeats in 3 different endpoint handlers as well. Should we have a separate method to return `installableRules`?